sympy/matrices/tests/test_matrixbase.py

```diff
@@ -1765,6 +1765,51 @@ def test_inverse():
     assert all(type(m.inv(s)) is cls for s in 'GE ADJ LU CH LDL QR'.split())
 
 
+def test_inverse_symbolic_float_issue_26821():
+    Tau, Tau_syn_in, Tau_syn_ex, C_m, Tau_syn_gap = symbols("Tau Tau_syn_in Tau_syn_ex C_m Tau_syn_gap")
+    __h = symbols("__h")
+
+    M = Matrix([
+        [0,0,0,0,0,(1.0*Tau*__h-1.0*Tau_syn_in*__h)/(2.0*Tau-1.0*Tau_syn_in),-1.0*Tau*Tau_syn_in/(2.0*Tau-1.0*Tau_syn_in)],
+        [0,0,0,0,0,(-1.0*Tau*__h+1.0*Tau_syn_in*__h)/(2.0*Tau*Tau_syn_in-1.0*Tau_syn_in**2),1.0],
+        [0,(1.0*Tau*__h-1.0*Tau_syn_ex*__h)/(2.0*Tau-1.0*Tau_syn_ex),-1.0*Tau*Tau_syn_ex/(2.0*Tau-1.0*Tau_syn_ex),0,0,0,0],
+        [0,(-1.0*Tau*__h+1.0*Tau_syn_ex*__h)/(2.0*Tau*Tau_syn_ex-1.0*Tau_syn_ex**2),1.0,0,0,0,0],
+        [0,0,0,(1.0*Tau*__h-1.0*Tau_syn_gap*__h)/(2.0*Tau-1.0*Tau_syn_gap),-1.0*Tau*Tau_syn_gap/(2.0*Tau-1.0*Tau_syn_gap),0,0],
+        [0,0,0,(-1.0*Tau*__h+1.0*Tau_syn_gap*__h)/(2.0*Tau*Tau_syn_gap-1.0*Tau_syn_gap**2),1.0,0,0],
+        [1.0,-1.0*Tau*Tau_syn_ex*__h/(2.0*C_m*Tau-1.0*C_m*Tau_syn_ex),0,-1.0*Tau*Tau_syn_gap*__h/(2.0*C_m*Tau-1.0*C_m*Tau_syn_gap),0,-1.0*Tau*Tau_syn_in*__h/(2.0*C_m*Tau-1.0*C_m*Tau_syn_in),0]
+    ])
+
+    Mi = M.inv()
```

Comment:
But, calling ``Me = (t*M).exp()`` still does not terminate, so the original problem still exists.

Reply:
For this particular matrix that is not surprising although when I test with this PR it terminates quickly (a few seconds) with an error. The characteristic polynomial is a degree 7 irreducible polynomial with 280 symbolic terms. Any expression for the matrix exponential here would be enormous. The current algorithm for matrix exponentials depends on computing the eigenvalues which already fails for this matrix:
```python
In [1]:     Tau, Tau_syn_in, Tau_syn_ex, C_m, Tau_syn_gap = symbols("Tau Tau_syn_in Tau_syn_ex C_m Tau_syn_gap")
   ...:     __h = symbols("__h")
   ...:
   ...:     M = Matrix([
   ...:         [0,0,0,0,0,(1.0*Tau*__h-1.0*Tau_syn_in*__h)/(2.0*Tau-1.0*Tau_syn_in),-1.0*Tau*Tau_syn_in/(2.0*Tau-1.0*Tau_syn_in)],
   ...:         [0,0,0,0,0,(-1.0*Tau*__h+1.0*Tau_syn_in*__h)/(2.0*Tau*Tau_syn_in-1.0*Tau_syn_in**2),1.0],
   ...:         [0,(1.0*Tau*__h-1.0*Tau_syn_ex*__h)/(2.0*Tau-1.0*Tau_syn_ex),-1.0*Tau*Tau_syn_ex/(2.0*Tau-1.0*Tau_syn_ex),0,0,0,0],
   ...:         [0,(-1.0*Tau*__h+1.0*Tau_syn_ex*__h)/(2.0*Tau*Tau_syn_ex-1.0*Tau_syn_ex**2),1.0,0,0,0,0],
   ...:         [0,0,0,(1.0*Tau*__h-1.0*Tau_syn_gap*__h)/(2.0*Tau-1.0*Tau_syn_gap),-1.0*Tau*Tau_syn_gap/(2.0*Tau-1.0*Tau_syn_gap),0,0],
   ...:         [0,0,0,(-1.0*Tau*__h+1.0*Tau_syn_gap*__h)/(2.0*Tau*Tau_syn_gap-1.0*Tau_syn_gap**2),1.0,0,0],
   ...:         [1.0,-1.0*Tau*Tau_syn_ex*__h/(2.0*C_m*Tau-1.0*C_m*Tau_syn_ex),0,-1.0*Tau*Tau_syn_gap*__h/(2.0*C_m*Tau-1.0*C_m*Tau_syn_gap),0,-1.0*Tau*Tau_syn_in*__h/(
   ...: 2.0*C_m*Tau-1.0*C_m*Tau_syn_in),0]
   ...:     ])

In [2]: %time Me = (M*t).exp()
---------------------------------------------------------------------------
MatrixError: It is not always possible to express the eigenvalues of a matrix of size 5x5 or higher in radicals. We have CRootOf, but domains other than the rationals are not currently supported. If there are no symbols in the matrix, it should still be possible to compute numeric approximations of the eigenvalues using M.evalf().eigenvals() or M.charpoly().nroots().

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
...
NotImplementedError: Exponentiation is implemented only for matrices for which the Jordan normal form can be computed
```

This test is added because computing the inverse of this matrix arises as part of computing the matrix exponential for the simpler matrix that is shown in gh-26821 which is the issue about large floats.

For the matrix reported in gh-26821 we have:
```python
In [11]: from sympy import *
    ...:
    ...: # Define symbols
    ...: Tau, Tau_syn_in, Tau_syn_ex, C_m, Tau_syn_gap = symbols("Tau Tau_syn_in Tau_syn_ex C_m Tau_syn_gap")
    ...:
    ...: M = Matrix([
    ...:  [               0,           1.0,                0,             0,                 0,              0,      0],
    ...:  [-1/Tau_syn_in**2, -2/Tau_syn_in,                0,             0,                 0,              0,      0],
    ...:  [               0,             0,                0,           1.0,                 0,              0,      0],
    ...:  [               0,             0, -1/Tau_syn_ex**2, -2/Tau_syn_ex,                 0,              0,      0],
    ...:  [               0,             0,                0,             0,                 0,              1,      0],
    ...:  [               0,             0,                0,             0, -1/Tau_syn_gap**2, -2/Tau_syn_gap,      0],
    ...:  [           1/C_m,             0,            1/C_m,             0,             1/C_m,              0, -1/Tau]]) * Symbol("__h")
    ...:

In [12]: %time Me = M.exp()
CPU times: user 17 s, sys: 7.8 ms, total: 17 s
Wall time: 17.1 s

In [13]: Me.atoms(Float)
Out[13]: {-128.0, -64.0, -32.0, 1.0, 2.0, 16.0, 32.0, 64.0}
```
So it takes 17 seconds on this computer and does not have large floats. With master or 1.13.0 we instead have:
```python
In [2]: %time Me = M.exp()
CPU times: user 19.9 s, sys: 21.7 ms, total: 19.9 s
Wall time: 20 s

In [3]: max(Me.atoms(Float))
Out[3]: 1.14555615673900e+251
```
So it takes about the same time (20 seconds) but has very large floats. The improvement in this PR is about not having these large floats which is the issue that was reported in gh-26821.

With SymPy 1.12.1 we have:
```python
In [2]: %time Me = M.exp()
CPU times: user 4min 49s, sys: 138 ms, total: 4min 49s
Wall time: 4min 50s

In [3]: max(Me.atoms(Float))
Out[3]: 2.36118324143482e+21
```
So computing the exponential of this matrix in SymPy 1.12.1 takes 5 minutes and produces a result with unnecessarily large floats. In SymPy 1.13.0 this became 15x faster but produces even larger floats.

This PR eliminates the large float problem and does not introduce any slowdown for this case which is still 15x faster than in SymPy 1.12.1.

Note that this PR is not intended as a fix for gh-19887 which is about the performance of matrix exponential. I did want to fix that before releasing SymPy 1.13.0 because the other pieces are there now to make this faster (e.g. faster charpoly, faster inverse, faster nullspace) but the release was already delayed long enough (over a year).

This PR is intended to be a bugfix for matrix inverse returning large floats which is a regression in SymPy 1.13.0 that I want to fix in a bugfix 1.13.1. I am not going to try to fix gh-19887 as part of that bugfix release because it needs more time to think it through carefully. The proper fix for gh-19887 means using a completely different algorithm for matrix exponential (gh-21585). Other matrix operations have been made faster in SymPy 1.13.0 which improves the situation but the completely different algorithm using RootSum is really what is needed. Note that the RootSum method does not depend on computing the eigenvalues and can theoretically compute the matrix exponential of any matrix of any size with any type of expressions.

Reply:
Thank you for the detailed explanation! Then I am fully on board with this. Cheers!

Reply:
Thanks for reviewing!